rmw_dds_common/src/graph_cache.cpp

```diff
@@ -0,0 +1,966 @@
+// limitations under the License.
+
+#include "rmw_dds_common/graph_cache.hpp"
+
+#include <algorithm>
+#include <cassert>
+#include <cstring>
+#include <functional>
+#include <iterator>
+#include <map>
+#include <memory>
+#include <mutex>
+#include <ostream>
+#include <set>
+#include <sstream>
+#include <string>
+#include <tuple>
+#include <utility>
+#include <vector>
+
+#include "rcutils/strdup.h"
+
+#include "rmw/convert_rcutils_ret_to_rmw_ret.h"
+#include "rmw/error_handling.h"
+#include "rmw/sanity_checks.h"
+#include "rmw/topic_endpoint_info.h"
+#include "rmw/topic_endpoint_info_array.h"
+
+#include "rmw_dds_common/gid_utils.hpp"
+
+using rmw_dds_common::GraphCache;
+using rmw_dds_common::operator<<;
+
+static const char log_tag[] = "rmw_dds_common";
+
+bool
+GraphCache::add_writer(
+  const rmw_gid_t & gid,
+  const std::string & topic_name,
+  const std::string & type_name,
+  const rmw_gid_t & participant_gid,
+  const rmw_qos_profile_t & qos)
+{
+  std::lock_guard<std::mutex> guard(mutex_);
+  auto pair = data_writers_.emplace(
+    std::piecewise_construct,
+    std::forward_as_tuple(gid),
+    std::forward_as_tuple(topic_name, type_name, participant_gid, qos));
+  return pair.second;
+}
+
+bool
+GraphCache::add_reader(
+  const rmw_gid_t & gid,
+  const std::string & topic_name,
+  const std::string & type_name,
+  const rmw_gid_t & participant_gid,
+  const rmw_qos_profile_t & qos)
+{
+  std::lock_guard<std::mutex> guard(mutex_);
+  auto pair = data_readers_.emplace(
+    std::piecewise_construct,
+    std::forward_as_tuple(gid),
+    std::forward_as_tuple(topic_name, type_name, participant_gid, qos));
+  return pair.second;
+}
+
+bool
+GraphCache::add_entity(
+  const rmw_gid_t & gid,
+  const std::string & topic_name,
+  const std::string & type_name,
+  const rmw_gid_t & participant_gid,
+  const rmw_qos_profile_t & qos,
+  bool is_reader)
+{
+  if (is_reader) {
+    return this->add_reader(
+      gid,
+      topic_name,
+      type_name,
+      participant_gid,
+      qos);
+  }
+  return this->add_writer(
+    gid,
+    topic_name,
+    type_name,
+    participant_gid,
+    qos);
+}
+
+bool
+GraphCache::remove_writer(const rmw_gid_t & gid)
+{
+  std::lock_guard<std::mutex> guard(mutex_);
+  return data_writers_.erase(gid) > 0;
+}
+
+bool
+GraphCache::remove_reader(const rmw_gid_t & gid)
+{
+  std::lock_guard<std::mutex> guard(mutex_);
+  return data_readers_.erase(gid) > 0;
+}
+
+bool
+GraphCache::remove_entity(const rmw_gid_t & gid, bool is_reader)
+{
+  if (is_reader) {
+    return this->remove_reader(gid);
+  }
+  return this->remove_writer(gid);
+}
+
+void
+GraphCache::update_participant_entities(const rmw_dds_common::msg::ParticipantEntitiesInfo & msg)
+{
+  std::lock_guard<std::mutex> guard(mutex_);
+  rmw_gid_t gid;
+  rmw_dds_common::convert_msg_to_gid(&msg.gid, &gid);
+  if (0u != msg.node_entities_info_seq.size()) {
+    participants_[gid] = msg.node_entities_info_seq;
+  } else {
+    participants_.erase(gid);
```

Comment:
Is it not possible for a participant to exist without any nodes? Is that case considered here?

Reply:
Why not just use `remove_participant()` below, instead of this sort of hidden behavior (docstring for this function doesn't mention it at all)?

Reply:
Plus you can be in this state right after `add_participant()`...

Reply:
> Is it not possible for a participant to exist without any nodes? Is that case considered here?

That makes sense, I can remove that.

> Why not just use remove_participant() below, instead of this sort of hidden behavior (docstring for this function doesn't mention it at all)?

`remove_participant` is also used (see https://github.com/ivanpauno/rmw_dds_common/pull/4#discussion_r376090067).

> Plus you can be in this state right after add_participant()...

`add_participant` is never used for a remote participant, a `Participant` is added when the first status message is received.

Reply:
solved